# backtesting

A quick test of the portfolio methodology via backtesting, with some helpful functions. 

## Imports

In [29]:
import pandas as pd
import quandl as ql
import numpy as np
import matplotlib.pyplot as plt
import cvxopt as opt
from cvxopt import blas, solvers
import warnings

# Make Chart dimensions waaaay bigger, some ptplot option
%matplotlib inline 
# Less busy convex solver logs
solvers.options['show_progress'] = False

## Data Retrieval

Retrieving high fidelity data from (Kaggle.)[https://www.kaggle.com/jessevent/all-crypto-currencies/data] Would prefer Quandl, but costs money.

In [6]:
df = pd.read_csv('data/crypto.csv')

In [63]:
symbols = df.symbol.unique()
slugs = df.slug.unique()

In [67]:
df.loc[df.symbol == 'STEEM'].tail(100)

slug symbol   name        date  ranknow      open     high       low  \
18093  steem  STEEM  Steem  2017-09-25       27  1.070000  1.17000  1.040000   
18094  steem  STEEM  Steem  2017-09-26       27  1.110000  1.12000  1.070000   
18095  steem  STEEM  Steem  2017-09-27       27  1.080000  1.26000  1.080000   
18096  steem  STEEM  Steem  2017-09-28       27  1.220000  1.49000  1.200000   
18097  steem  STEEM  Steem  2017-09-29       27  1.440000  1.64000  1.260000   
18098  steem  STEEM  Steem  2017-09-30       27  1.640000  1.68000  1.430000   
18099  steem  STEEM  Steem  2017-10-01       27  1.460000  1.52000  1.410000   
18100  steem  STEEM  Steem  2017-10-02       27  1.390000  1.43000  1.290000   
18101  steem  STEEM  Steem  2017-10-03       27  1.320000  1.33000  1.170000   
18102  steem  STEEM  Steem  2017-10-04       27  1.220000  1.23000  1.130000   
18103  steem  STEEM  Steem  2017-10-05       27  1.180000  1.22000  1.090000   
18104  steem  STEEM  Steem  2017-10-06       27  1.170000  1.23000  1.170000   
18105  steem  STEEM  Steem  2017-10-07       27  1.210000  1.26000  1.170000   
18106  steem  STEEM  Steem  2017-10-08       27  1.220000  1.27000  1.080000   
18107  steem  STEEM  Steem  2017-10-09       27  1.120000  1.15000  0.958003   
18108  steem  STEEM  Steem  2017-10-10       27  1.070000  1.13000  1.020000   
18109  steem  STEEM  Steem  2017-10-11       27  1.030000  1.12000  1.030000   
18110  steem  STEEM  Steem  2017-10-12       27  1.100000  1.12000  1.010000   
18111  steem  STEEM  Steem  2017-10-13       27  1.040000  1.04000  0.967398   
18112  steem  STEEM  Steem  2017-10-14       27  1.030000  1.18000  1.010000   
18113  steem  STEEM  Steem  2017-10-15       27  1.070000  1.10000  0.998666   
18114  steem  STEEM  Steem  2017-10-16       27  1.070000  1.08000  1.010000   
18115  steem  STEEM  Steem  2017-10-17       27  1.050000  1.09000  1.010000   
18116  steem  STEEM  Steem  2017-10-18       27  1.060000  1.11000  0.971589   
18117  steem  STEEM  Steem  2017-10-19       27  1.070000  1.09000  1.000000   
18118  steem  STEEM  Steem  2017-10-20       27  1.090000  1.13000  0.993585   
18119  steem  STEEM  Steem  2017-10-21       27  0.999598  1.02000  0.893548   
18120  steem  STEEM  Steem  2017-10-22       27  0.972041  1.03000  0.939994   
18121  steem  STEEM  Steem  2017-10-23       27  0.974869  0.98807  0.880831   
18122  steem  STEEM  Steem  2017-10-24       27  0.930480  1.07000  0.898789   
...      ...    ...    ...         ...      ...       ...      ...       ...   
18163  steem  STEEM  Steem  2017-12-04       27  1.220000  1.46000  1.190000   
18164  steem  STEEM  Steem  2017-12-05       27  1.430000  1.62000  1.420000   
18165  steem  STEEM  Steem  2017-12-06       27  1.450000  1.48000  1.260000   
18166  steem  STEEM  Steem  2017-12-07       27  1.290000  2.15000  1.270000   
18167  steem  STEEM  Steem  2017-12-08       27  2.140000  2.16000  1.610000   
18168  steem  STEEM  Steem  2017-12-09       27  1.900000  1.94000  1.540000   
18169  steem  STEEM  Steem  2017-12-10       27  1.670000  1.92000  1.310000   
18170  steem  STEEM  Steem  2017-12-11       27  1.680000  1.79000  1.680000   
18171  steem  STEEM  Steem  2017-12-12       27  1.720000  1.87000  1.660000   
18172  steem  STEEM  Steem  2017-12-13       27  1.860000  1.88000  1.690000   
18173  steem  STEEM  Steem  2017-12-14       27  1.760000  2.10000  1.760000   
18174  steem  STEEM  Steem  2017-12-15       27  2.040000  2.08000  1.800000   
18175  steem  STEEM  Steem  2017-12-16       27  1.970000  2.26000  1.970000   
18176  steem  STEEM  Steem  2017-12-17       27  2.110000  2.42000  2.100000   
18177  steem  STEEM  Steem  2017-12-18       27  2.330000  2.94000  2.210000   
18178  steem  STEEM  Steem  2017-12-19       27  2.870000  3.37000  2.630000   
18179  steem  STEEM  Steem  2017-12-20       27  2.880000  3.70000  2.430000   
18180  steem  STEEM  Steem  2017-12-21       27  3.440000  4.17000  3.310000   
18181  s

In [69]:
required_symbols = ['BTC','ETH','DASH','XMR','QTUM','WAVES','LSK','OMG','STEEM','REP', 'GNT','SC','DGD','FCT','MAID','PAY','SNT','BAT','ZRX','BNT','CVC','STORJ','SINGLS','ANT','RLC','TRST','DNT']

BLX: 

* Bitcoin (BTC)	Value	15.00 %	7.04 %
* Ethereum (ETH)	Platform	14.17 %	14.49 %
* Dash (DASH)	Anonymity	12.89 %	8.36 %
* Monero (XMR)	Anonymity	12.68 %	8.48 %
* Qtum (QTUM)	Platform	6.90 %	10.82 %
* Waves (WAVES)	Platform	6.50 %	3.22 %
* Lisk (LSK)	Platform	6.21 %	9.20 %
* OmiseGO (OMG)	Decentralized Exchange and Payments Platform	5.62 %	6.55 %
* Steem (STEEM)	Media	2.41 %	3.54 %
* Augur (REP)	Prediction	2.02 %	2.66 %
* Golem (GNT)	Decentralized Computing	1.75 %	2.04 %
* Siacoin (SC)	Storage	1.60 %	5.07 %
* DigixDAO (DGD)	Value Gold	1.47 %	1.22 %
* Factom (FCT)	Audit	1.33 %	1.52 %
* MaidSafeCoin (MAID)	Storage	1.27 %	1.19 %
* TenX (PAY)	Debit Card	1.27 %	1.04 %
* Status (SNT)	Smartphone Ethereum Gateway	1.20 %	3.56 %
* Basic Attention Token (BAT)	Advertisement	1.18 %	2.10 %
* 0x (ZRX)	Decentralized Exchange Protocol	0.86 %	2.90 %
* Bancor (BNT)	Financial Derivative	0.78 %	1.09 %
* Civic (CVC)	Identity Management	0.72 %	0.90 %
* Storj (STORJ)	Storage	0.54 %	0.72 %
* SingularDTV (SNGLS)	Media / TV	0.48 %	0.55 %
* Aragon (ANT)	Enterprise Management	0.43 %	0.54 %
* iExec (RLC)	Decentralized Computing	0.33 %	0.49 %
* WeTrust (TRST)	Lending	0.24 %	0.25 %
* District0x (DNT)	Decentralized infrastructure	0.15 %	0.46 %

In [65]:
print(slugs)
print(symbols)

['bitcoin' 'ripple' 'ethereum' ..., 'aidos-kuneen' 'casinocoin'
 'bitcoin-unlimited']
['BTC' 'XRP' 'ETH' ..., 'ADK' 'CSC' 'BTU']


In [66]:
[print(i) for i in slugs]
[print(i) for i in symbols]

bitcoin
ripple
ethereum
bitcoin-cash
cardano
litecoin
stellar
nem
iota
dash
monero
neo
tron
eos
bitcoin-gold
qtum
raiblocks
ethereum-classic
bitconnect
icon
lisk
bitshares
verge
omisego
status
ardor
steem
zcash
populous
stratis
tether
waves
bytecoin-bcn
hshare
dogecoin
komodo
siacoin
golem-network-tokens
binance-coin
augur
vechain
veritaseum
ark
kucoin-shares
decred
digibyte
salt
pivx
dragonchain
nxt
factom
aion
monacoin
byteball
basic-attention-token
maidsafecoin
request-network
zcoin
enigma-project
power-ledger
bitcoindark
civic
0x
reddcoin
tenx
experience-points
electroneum
kyber-network
wax
syscoin
santiment
funfair
ethos
bytom
rchain
digixdao
qash
kin
xtrabytes
aeternity
gas
vertcoin
gnosis-gno
gamecredits
storj
iconomi
quantstamp
dent
substratum
aelf
gxshares
skycoin
ubiq
raiden-network-token
walton
centra
decentraland
bancor
storm
bitbay
blocknet
chainlink
edgeless
smartcash
poet
red-pulse
time-new-bank
iot-chain
triggers
revain
nav-coin
cryptonex
monaco
einsteinium
bridgecoin
r

beavercoin
blackstar
bitquark
dreamcoin
jewels
mindcoin
asiadigicoin
cryptoworldx-token
litecred
morningstar-payments
billarycoin
vip-tokens
pulse
warp
cashcoin
quebecoin
uro
ego
orlycoin
dappster
dpay
crtcoin
plncoin
zonecoin
mtmgaming
bios-crypto
playercoin
impulsecoin
zayedcoin
bowscoin
jobscoin
ponzicoin
tagrcoin
genstake
high-voltage
wild-beast-block
arbit
letitride
posex
agrolifecoin
speedcash
xonecoin
osmiumcoin
gbcgoldcoin
antilitecoin
kingn-coin
ibank
sydpak
cabbage
cryptoescudo
futurexe
p7coin
ccminer
corethum
selfiecoin
biobar
nodecoin
qibuck-asset
socialcoin-socc
ulatech
save-and-gain
concoin
lex4all
digital-credits
pizzacoin
caliphcoin
applecoin-apw
atmcoin
genaro-network
nebulas-token
bibox-token
smartmesh
entcash
bitcoin-diamond
energo
dimcoin
medishares
naga
html-coin
prochain
segwit2x
qlink
comsa-eth
hyper-pay
micromoney
olympus-labs
sophiatx
dew
tezos
bitclave
bitcoinx
clubcoin
homeblockcoin
comsa-xem
qbao
super-bitcoin
fargocoin
embercoin
lltoken
poly-ai
cappasity
in

[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,

In [86]:
coins = []

[print(name) if (len(df.loc[df.symbol == name].tail(365))>0) else print('FUCK',name) for name in required_symbols]

BTC
ETH
DASH
XMR
QTUM
WAVES
LSK
OMG
STEEM
REP
GNT
SC
DGD
FCT
MAID
PAY
SNT
BAT
ZRX
BNT
CVC
STORJ
FUCK SINGLS
ANT
RLC
TRST
DNT


[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None]

In [75]:
df.loc[df.symbol == 'QTUM'].tail(100)

slug symbol  name        date  ranknow   open   high    low  close  \
12348  qtum   QTUM  Qtum  2017-09-25       16   8.03   9.60   8.03   9.36   
12349  qtum   QTUM  Qtum  2017-09-26       16   9.38  10.30   9.17   9.34   
12350  qtum   QTUM  Qtum  2017-09-27       16   9.37   9.90   9.35   9.90   
12351  qtum   QTUM  Qtum  2017-09-28       16   9.95  10.10   9.06   9.93   
12352  qtum   QTUM  Qtum  2017-09-29       16   9.96  10.00   9.15   9.77   
12353  qtum   QTUM  Qtum  2017-09-30       16   9.77  11.65   9.75  11.59   
12354  qtum   QTUM  Qtum  2017-10-01       16  11.59  12.61  11.58  12.34   
12355  qtum   QTUM  Qtum  2017-10-02       16  12.36  12.38  11.43  11.88   
12356  qtum   QTUM  Qtum  2017-10-03       16  11.91  12.12  10.78  12.12   
12357  qtum   QTUM  Qtum  2017-10-04       16  12.16  12.28  11.51  11.63   
12358  qtum   QTUM  Qtum  2017-10-05       16  11.60  11.98  10.82  11.86   
12359  qtum   QTUM  Qtum  2017-10-06       16  11.87  11.91  11.50  11.69   
12360  qtum   QTUM  Qtum  2017-10-07       16  11.67  11.94  11.51  11.94   
12361  qtum   QTUM  Qtum  2017-10-08       16  11.96  12.11  11.14  11.14   
12362  qtum   QTUM  Qtum  2017-10-09       16  11.14  12.49  10.96  12.07   
12363  qtum   QTUM  Qtum  2017-10-10       16  12.04  12.54  11.55  12.08   
12364  qtum   QTUM  Qtum  2017-10-11       16  12.04  12.29  11.87  12.13   
12365  qtum   QTUM  Qtum  2017-10-12       16  12.13  12.21  11.29  11.29   
12366  qtum   QTUM  Qtum  2017-10-13       16  11.30  11.30  10.31  11.25   
12367  qtum   QTUM  Qtum  2017-10-14       16  11.26  11.45  10.62  10.63   
12368  qtum   QTUM  Qtum  2017-10-15       16  10.64  10.66  10.09  10.44   
12369  qtum   QTUM  Qtum  2017-10-16       16  10.46  11.65  10.23  11.49   
12370  qtum   QTUM  Qtum  2017-10-17       16  11.48  12.27  11.03  12.05   
12371  qtum   QTUM  Qtum  2017-10-18       16  12.04  12.04  10.78  11.49   
12372  qtum   QTUM  Qtum  2017-10-19       16  11.46  12.22  11.35  12.07   
12373  qtum   QTUM  Qtum  2017-10-20       16  12.10  15.12  11.19  11.30   
12374  qtum   QTUM  Qtum  2017-10-21       16  11.25  11.35  10.50  10.91   
12375  qtum   QTUM  Qtum  2017-10-22       16  10.97  11.06  10.56  10.63   
12376  qtum   QTUM  Qtum  2017-10-23       16  10.59  10.60   9.75   9.96   
12377  qtum   QTUM  Qtum  2017-10-24       16   9.96  11.46   9.52  10.76   
...     ...    ...   ...         ...      ...    ...    ...    ...    ...   
12418  qtum   QTUM  Qtum  2017-12-04       16  13.15  14.00  12.68  13.62   
12419  qtum   QTUM  Qtum  2017-12-05       16  13.69  13.95  13.08  13.08   
12420  qtum   QTUM  Qtum  2017-12-06       16  13.12  13.14  11.86  11.96   
12421  qtum   QTUM  Qtum  2017-12-07       16  11.95  14.93  11.67  12.56   
12422  qtum   QTUM  Qtum  2017-12-08       16  12.53  13.06  11.99  12.79   
12423  qtum   QTUM  Qtum  2017-12-09       16  12.77  12.77  11.54  12.08   
12424  qtum   QTUM  Qtum  2017-12-10       16  12.09  12.09  10.74  11.05   
12425  qtum   QTUM  Qtum  2017-12-11       16  11.02  14.60  11.00  13.55   
12426  qtum   QTUM  Qtum  2017-12-12       16  13.57  15.12  13.28  14.76   
12427  qtum   QTUM  Qtum  2017-12-13       16  14.81  19.29  14.80  17.31   
12428  qtum   QTUM  Qtum  2017-12-14       16  17.16  20.63  17.02  20.48   
12429  qtum   QTUM  Qtum  2017-12-15       16  20.68  32.65  17.69  30.54   
12430  qtum   QTUM  Qtum  2017-12-16       16  30.15  31.53  28.48  29.00   
12431  qtum   QTUM  Qtum  2017-12-17       16  29.19  30.30  28.63  29.62   
12432  qtum   QTUM  Qtum  2017-12-18       16  29.66  56.06  27.84  51.85   
12433  qtum   QTUM  Qtum  2017-12-19       16  51.96  77.75  51.30  67.75   
12434  qtum   QTUM  Qtum  2017-12-20       16  67.72  69.07  57.71  63.64   
12435  qtum   QTUM  Qtum  2017-12-21       16  63.82  69.24  56.87  62.23   
12436  qtum   QTUM  Qtum  2017-12-22       16  63.27  63.27  36.35  49.48   
12437  qtum   QTUM  Qtum  2017-12-23       16  50.84  60.20  50.71  57.26   
124

In [43]:
daily = pd.date_range('1/3/2016', end='1/3/2018', freq='D')
weekly = pd.date_range('1/3/2016', end='1/3/2018', freq='W')
print(daily)
print(weekly)

DatetimeIndex(['2016-01-03', '2016-01-04', '2016-01-05', '2016-01-06',
               '2016-01-07', '2016-01-08', '2016-01-09', '2016-01-10',
               '2016-01-11', '2016-01-12',
               ...
               '2017-12-25', '2017-12-26', '2017-12-27', '2017-12-28',
               '2017-12-29', '2017-12-30', '2017-12-31', '2018-01-01',
               '2018-01-02', '2018-01-03'],
              dtype='datetime64[ns]', length=732, freq='D')
DatetimeIndex(['2016-01-03', '2016-01-10', '2016-01-17', '2016-01-24',
               '2016-01-31', '2016-02-07', '2016-02-14', '2016-02-21',
               '2016-02-28', '2016-03-06',
               ...
               '2017-10-29', '2017-11-05', '2017-11-12', '2017-11-19',
               '2017-11-26', '2017-12-03', '2017-12-10', '2017-12-17',
               '2017-12-24', '2017-12-31'],
              dtype='datetime64[ns]', length=105, freq='W-SUN')


In [47]:
opt.matrix(np.cov(eth['open'], btc['open']))

<2x2 matrix, tc='d'>

In [33]:
np.asmatrix(eth['open'])

matrix([[  2.83000000e+00,   2.79000000e+00,   7.06136000e-01,
           7.13989000e-01,   7.08087000e-01,   1.06000000e+00,
           1.22000000e+00,   1.81000000e+00,   1.80000000e+00,
           1.68000000e+00,   1.58000000e+00,   1.22000000e+00,
           1.17000000e+00,   1.25000000e+00,   1.48000000e+00,
           1.40000000e+00,   1.38000000e+00,   1.35000000e+00,
           1.23000000e+00,   1.13000000e+00,   1.17000000e+00,
           1.15000000e+00,   1.19000000e+00,   1.18000000e+00,
           1.32000000e+00,   1.35000000e+00,   1.35000000e+00,
           1.30000000e+00,   1.26000000e+00,   1.28000000e+00,
           1.34000000e+00,   1.30000000e+00,   1.25000000e+00,
           1.24000000e+00,   1.21000000e+00,   1.16000000e+00,
           9.87228000e-01,   1.04000000e+00,   9.40566000e-01,
           8.75189000e-01,   9.41977000e-01,   9.06865000e-01,
           8.74574000e-01,   8.49603000e-01,   8.81191000e-01,
           9.41496000e-01,   9.19337000e-01,   9.035600

In [36]:
## NUMBER OF ASSETS
n_assets = 4

## NUMBER OF OBSERVATIONS
n_obs = 1000

return_vec = np.random.randn(n_assets, n_obs)

In [38]:
def rand_weights(n):
    ''' Produces n random weights that sum to 1 '''
    k = np.random.rand(n)
    return k / sum(k)

def random_portfolio(returns):
    ''' 
    Returns the mean and standard deviation of returns for a random portfolio
    '''

    p = np.asmatrix(np.mean(returns, axis=1))
    w = np.asmatrix(rand_weights(returns.shape[0]))
    C = np.asmatrix(np.cov(returns))
    
    mu = w * p.T
    sigma = np.sqrt(w * C * w.T)
    
    # This recursion reduces outliers to keep plots pretty
    if sigma > 2:
        return random_portfolio(returns)
    return mu, sigma

random_portfolio(return_vec)

(matrix([[ 0.00948996]]), matrix([[ 0.51000164]]))

### Todo:
* Pick the crypto to put in the portfolio (maybe top 20 by market cap in the data?)
* Generate a weight vector every week in the data, simulating real life (trained on the past year to date? All historical to current? Last week of data only?) 
* Given this weight vector, simulate a portfolio of that position for the week, and calculate expected performance.